analyses/cms-open-data-ttbar/utils.py

```diff
@@ -24,10 +52,95 @@ def update_to(b=1, bsize=1, tsize=None):
 
     return update_to
 
-def cache_files(file_paths: list, cache_dir: str, remote_prefix: str):
+
+def _cache_files(file_paths: list, cache_dir: str, remote_prefix: str):
     for url in file_paths:
-        out_path = Path(cache_dir) / url.removeprefix(remote_prefix).lstrip('/')
+        out_path = Path(cache_dir) / url.removeprefix(remote_prefix).lstrip("/")
         out_path.parent.mkdir(parents=True, exist_ok=True)
         if not out_path.exists():
-            with tqdm(unit='B', unit_scale=True, unit_divisor=1024, miniters=1, desc=out_path.name) as t:
+            with tqdm(
+                unit="B", unit_scale=True, unit_divisor=1024, miniters=1, desc=out_path.name
+            ) as t:
                 urlretrieve(url, out_path.absolute(), reporthook=_tqdm_urlretrieve_hook(t))
+
+
+def retrieve_inputs(
+    max_files_per_sample: Optional[int],
+    remote_data_prefix: Optional[str],
+    data_cache: Optional[str],
+    input_json: Path = Path("ntuples.json"),
+) -> list[AGCInput]:
+    """Return a dictionary of file paths and a corresponding dictionary of event counts.
+    Both are 2-level dictionaries: there is a dictionary per process per variation.
+    Each files[process][variation] corresponds to a list of input files.
+    nevts[process][variation] is the total number of events for that sample.
+    """
+    with open(input_json) as f:
+        input_spec = json.load(f)
+
+    input_files: list[AGCInput] = []
+
+    for process, process_spec in input_spec.items():
+        if process == "data":
+            continue  # skip data
+
+        for variation, sample_info in process_spec.items():
+            sample_info = sample_info["files"]  # this is now a list of (filename, nevents) tuples
+
+            if max_files_per_sample is not None:
+                sample_info = sample_info[:max_files_per_sample]
+
+            file_paths = [f["path"] for f in sample_info]
+            prefix = "https://xrootd-local.unl.edu:1094//store/user/AGC"
+            assert all(f.startswith(prefix) for f in file_paths)
+
+            if remote_data_prefix is not None:
+                old_prefix, prefix = prefix, remote_data_prefix
+                file_paths = [f.replace(old_prefix, prefix) for f in file_paths]
+
+            if data_cache is not None:
+                _cache_files(file_paths, data_cache, prefix)
+                old_prefix, prefix = prefix, str(Path(data_cache).absolute())
+                file_paths = [f.replace(old_prefix, prefix) for f in file_paths]
+
+            nevents = sum(f["nevts"] for f in sample_info)
+            input_files.append(AGCInput(file_paths, process, variation, nevents))
+
+    return input_files
+
+
+def postprocess_results(results: list[AGCResult]):
+    """Extract TH1D objects from list of RDF's ResultPtrs and RResultMaps.
+    The function also gives appropriate names to each varied histogram and slices them and rebins them as needed.
```

Comment:
I am missing the part of the function that slices/rebins the histograms

Reply:
ah, thank you, it doesn't anymore (see third commit of this PR). I added a commit that fixes the docs.